### PR TITLE
[Ray Job] Add support for Windows OS to JobSupervisor

### PR DESCRIPTION
**Pull request by jbedorf**
## Why are these changes needed?

The existing Ray Job submit/launch code uses a number of Unix only API calls that break the submission/status flow when using Ray Job.  Specifically the following items are broken:

* Submission, the job/work is correctly launched but due to the call to `pgid` the `JobSupervisor` will crash as that function is not available on Windows.
* The removal of child processes is not working due to `subprocess.kill()` only killing the shell process (cmd.exe) but not the actual submitted work.

This PR fixes the above issues by using win32 APIs to setup the connection between the JobSupervisor class and the launched workers. This results in the behaviour to be the same for Unix and Windows based executions.

The one requirement is that the `pywin32` package is installed. If it is not installed a warning will be printed and the code will function as before, but with the addition that the JobSupervisor will not crash. The thing that will be missing is the correct shutdown of the child process.

 
## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by jbedorf** [comment]
@jiaodong  Can you help find someone who can review this? (Tagging you as there was a TODO with your name in the code, so I assume you know at least who's responsible for these files.) Thanks!.

**Comment by jbedorf** [comment]
@alanwguo  given that you edited the modified file recently, can you help find someone to review this and move it along? I've just updated it to the latest master.  Thanks!

**Review by architkulkarni — COMMENTED** [review]
Thanks for the fix!  I'm not too familiar with Windows so I'll ask @mattip to review this as well, but the change looks reasonable.

Can you please add a test for this, perhaps in `test_job_manager.py`?

@richardliaw Any thoughts on whether we should add `pywin32` to `requirements.txt` as a conditional install for Windows?

File: dashboard/modules/job/job_manager.py [hunk 1]
```diff
@@ -180,18 +203,45 @@ def _exec_entrypoint(self, logs_path: str) -> subprocess.Popen:
...
+                # alive. If it fails, SIGKILL the child process group and exit
+                subprocess.Popen(
+                    f"while kill -s 0 {parent_pid}; do sleep 1; done; kill -9 -{child_pgid}",  # noqa: E501
+                    shell=True,
+                    # Suppress output
+                    stdout=subprocess.DEVNULL,
+                    stderr=subprocess.DEVNULL,
+                )
+
+            if sys.platform == "win32" and win32api:
```
Comment (mattip) [inline dashboard/modules/job/job_manager.py:221]:
```suggestion
            elif sys.platform == "win32" and win32api:
```

File: dashboard/modules/job/job_manager.py [hunk 2]
```diff
@@ -180,18 +203,45 @@ def _exec_entrypoint(self, logs_path: str) -> subprocess.Popen:
             parent_pid = os.getpid()
             # Create new pgid with new subprocess to execute driver command
             child_pid = child_process.pid
```
Comment (mattip) [inline dashboard/modules/job/job_manager.py:205]:
```suggestion
            child_pid = child_process.pid
            # Create new pgid with new subprocess to execute driver command
```

**Review by mattip — COMMENTED** [review]
I have some tiny review nits.

There are tests in dashboard/modules/job/tests/test_job_manager.py. How are they run on linux? I see the regular CI runs a bazel test in dashboard, and this line is missing from the pipelie.windows.yml script
https://github.com/ray-project/ray/blob/bbd13ddc3361bbdd15fa93fe6ef4f30eec68b5c0/.buildkite/pipeline.yml#L241

**Comment by jbedorf** [comment]
I had a look at the tests, but that would be a whole project on it's own. The current tests are not run on Windows and the tests are written with the assumption it is executed on Linux. A large number of them fail for a variety of reasons. 

Example of something that is easy to fix:
Windows line endings that are different than Linux's but the tests all assume them to be '\n' instead of '\r\n'. That can easily be fixed by adding `.strip()`. 

Example of something that is harder to fix:
Due to the assumption that the executing shell is 'bash', whereas on Windows this is 'cmd' and as such the more [complicated bash scripts](https://github.com/ray-project/ray/blob/910ace838207202cf8dbbb9a64ebc19352b82c1b/dashboard/modules/job/tests/test_job_manager.py#L100) that are used in the test will not work as 'cmd' does not know the syntax. The one linked could be replaced by something along the lines as the snippet below that would allow it to work on Linux and Windows.

And then finally there are some additional issues with the execution itself, I think environment variables are not properly passed to the launched job (I didn't look into the why of this yet), but that causes some of the tests to fail as well.

Python version of the bash script in `_run_hanging_command`
```
    pid_file2 = str(pid_file).replace('\\', '/')
    tmp_file2 = str(tmp_file).replace('\\', '/')

    script = f"""
import os, sys, time 

with open("{pid_file2}", "w") as f:
    f.write(str(os.getpid()))
while not os.path.exists("{tmp_file2}"):
    print("Waiting...")
    time.sleep(1)
    """

    with open(script_file, "w", encoding="utf-8") as f:
        f.write(script)
    wait_for_file_cmd = f"python {script_file}"
```

**Comment by jbedorf** [comment]
@architkulkarni I've applied the suggest code changes, and with respect to the tests see my comment above. What would you like me to do to get this PR to move forward?

**Comment by architkulkarni** [comment]
Thanks for the investigation, I'll open an issue to track running these tests on Windows.

**Review by architkulkarni — APPROVED** [review]

**Comment by architkulkarni** [comment]
linkcheck failure unrelated: (tune/api_docs/suggestion: line  217) broken    https://app.sigopt.com/docs/overview/create - 404 Client Error: Not Found for url: https://docs.sigopt.com/experiments

**Comment by architkulkarni** [comment]
@simon-mo can you merge this when you get a chance?

**Pull request merged**